File: FindDiagonalOrder.java
package S30.exercises;

// Time Complexity : O(m * n)
// Space Complexity : O(1)
// Did this code successfully run on LeetCode : Yes
// Any problem you faced while coding this : No

// Approach: loop through the all the elements of the matrix and use a boolean direction flag and set it to true.
// Start each cell  data in the result array. While traversing up decrease the row index and increase the column index
// and make sure the column index is columns length, then change the flag value to false.
// While traversing down increase the row index and decrease the column index and make sure the row index is in boundries
// rows length and then make te boolean flag to true

public class FindDiagonalOrder {

    public int[] findDiagonalOrder(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        int size = m * n;
        int[] result = new int[size];
        boolean traverseUp = true;
        int idx = 0, i = 0, j = 0;
        while(idx < size){
            result[idx] = matrix[i][j]; // store first cell info.
            idx ++;
            if(traverseUp){  // traverse Up
                if(i==0 && j != n-1){
                    j++;
                    traverseUp = false;
                }else if(j== n-1){
                    i++;
                    traverseUp = false;
                }else{
                    i--; j++;
                }
            }else{
                //traverse down
                if(j==0 && i != m-1){
                    i++;
                    traverseUp = true;
                }else if(i== m-1){
                    j++;
                    traverseUp = true;
                }else{
                    j--; i++;
                }
            }
        }
       return result;
    }
}

File: findDiagonalOrder.js
// Time Complexity : O(m*n) where m*n is mat length
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {number[][]} mat
 * @return {number[]}
 */
 var findDiagonalOrder = function(mat) {
    if (mat === null) return [];
    let result = [];
    let m = mat.length;
    let n = mat[0].length;
    let i = 0, j = 0, index = 0;
    let dir = true;
    while(index < m*n) {
        result[index] = mat[i][j];
        index ++;
        //up dir in mat
        if(dir) {
            if(j == n-1) {
                dir = false;
                i++;
            } else if(i == 0) {
                dir = false;
                j++;
            } else {
                i--;
                j++;
            }
        } 
        //down dir in mat
        else {
             if(i == m-1) {
                dir = true;
                j++;
            } else if(j == 0) {
                dir = true;
                i++;
            } else {
                i++;
                j--;
            }
        }
    }
    return result;
};